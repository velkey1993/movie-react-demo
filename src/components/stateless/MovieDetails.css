#movie-details-container {
    position: relative;
    background: none;
    margin-top: 30px;
    border-radius: 0;
}

.jumbotron #movie-details-container-background {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-image: linear-gradient(
        rgba(30, 30, 30, 0.95),
        rgba(30, 30, 30, 0.95)
    ),
    url("../../assets/top-component-background.jpg");
    background-position: 0 25%;
    background-size: cover;
    background-repeat: no-repeat;

    z-index: -1;
}

#movie-details-container-page-name {
    color: #f65261;
}

#movie-details-container-search-button {
    color: #FF6565;
    background-color: rgba(85, 85, 85, 0);
    border: 0;
    font-size: 120%;
    -webkit-transform: scaleX(-1);
    transform: scaleX(-1);
}

#movie-details-container-data img {
    width: 220px;
    height: 325px;
}

#movie-details-container-data .container {
    padding-right: 0;
    padding-left: 0;
}

#movie-details-container-data-title {
    font-size: 300%;
    margin-bottom: 0
}

#movie-details-container-data-vote-average {
    border: 1px solid white;
    border-radius: 100%;
    font-size: 80%;
    padding: 5px;
    color: greenyellow
}
