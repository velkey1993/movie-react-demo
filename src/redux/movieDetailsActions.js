import * as movieService from '../service/MovieService';

export const FETCH_MOVIE_DETAILS_PENDING = 'FETCH_MOVIE_DETAILS_PENDING';
export const FETCH_MOVIE_DETAILS_ERROR = 'FETCH_MOVIE_DETAILS_ERROR';
export const FETCH_MOVIE_DETAILS_SUCCESS = 'FETCH_MOVIE_DETAILS_SUCCESS';

function pending() {
    return dispatch => dispatch({
        type: FETCH_MOVIE_DETAILS_PENDING,
    });
}

function handleError(error) {
    return dispatch => dispatch({
        type: FETCH_MOVIE_DETAILS_ERROR,
        error,
    });
}

export function fetchMovieDetailsSuccess(data) {
    return dispatch => dispatch({
        type: FETCH_MOVIE_DETAILS_SUCCESS,
        payload: data,
    });
}

export function fetchMovieDetails(id) {
    return (dispatch, getState) => {
        if (getState().movieDetails.pending) return Promise.reject(new Error('Pending action'));
        dispatch(pending());
        return movieService.readMovieDetails(id)
            .then((res) => {
                dispatch(fetchMovieDetailsSuccess(res.data));
            })
            .catch((error) => {
                dispatch(handleError(error?.response?.data.messages || error));
            });
    };
}
