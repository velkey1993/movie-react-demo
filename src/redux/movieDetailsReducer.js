import {
    FETCH_MOVIE_DETAILS_ERROR,
    FETCH_MOVIE_DETAILS_PENDING,
    FETCH_MOVIE_DETAILS_SUCCESS,
} from './movieDetailsActions';

const initialState = {
    pending: false,
    movieDetails: null,
    error: null,
};

const reducer = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_MOVIE_DETAILS_PENDING:
            return {
                ...state,
                pending: true,
            };

        case FETCH_MOVIE_DETAILS_ERROR:
            return {
                movieDetails: null,
                pending: false,
                error: action.error,
            };

        case FETCH_MOVIE_DETAILS_SUCCESS:
            return {
                ...state,
                movieDetails: action.payload,
                pending: false,
                error: null,
            };

        default:
            return state;
    }
};

export default reducer;
