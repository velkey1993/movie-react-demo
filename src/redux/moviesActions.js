import * as movieService from '../service/MovieService';

export const PENDING = 'PENDING';
export const ERROR = 'ERROR';
export const FETCH_MOVIES_SUCCESS = 'FETCH_MOVIES_SUCCESS';
export const FETCH_MOVIE_SUCCESS = 'FETCH_MOVIE_SUCCESS';
export const FETCH_MOVIES_PAGINATION_SUCCESS = 'FETCH_MOVIES_PAGINATION_SUCCESS';
export const ADD_MOVIE_SUCCESS = 'ADD_MOVIE_SUCCESS';
export const EDIT_MOVIE_SUCCESS = 'EDIT_MOVIE_SUCCESS';
export const DELETE_MOVIE_SUCCESS = 'DELETE_MOVIE_SUCCESS';

export const FETCH_BY_SEARCH = 'FETCH_BY_SEARCH';
export const FETCH_BY_ID = 'FETCH_BY_ID';

function pending(fetchType) {
    return dispatch => dispatch({
        type: PENDING,
        payload: fetchType,
    });
}

function handleError(error) {
    return dispatch => dispatch({
        type: ERROR,
        error,
    });
}

function fetchMoviesSuccess(data) {
    return dispatch => dispatch({
        type: FETCH_MOVIES_SUCCESS,
        payload: data,
    });
}

function fetchMovieSuccess(data) {
    return dispatch => dispatch({
        type: FETCH_MOVIE_SUCCESS,
        payload: data,
    });
}

function fetchMoviesPaginationSuccess(data) {
    return dispatch => dispatch({
        type: FETCH_MOVIES_PAGINATION_SUCCESS,
        payload: data,
    });
}

export function fetchMovies(sortBy, filter, search) {
    return (dispatch, getState) => {
        if (getState().movies.pending) return Promise.reject(new Error('Pending action'));
        dispatch(pending(FETCH_BY_SEARCH));
        return movieService.searchMovies(sortBy, filter, search)
            .then((res) => {
                dispatch(fetchMoviesSuccess(res.data));
            })
            .catch((e) => {
                const error = e?.response?.data.messages
                    ? new Error(e?.response?.data.messages) : e;
                dispatch(handleError(error.message));
                throw error;
            });
    };
}

export function fetchMovie(id) {
    return (dispatch, getState) => {
        if (getState().movies.pending) return Promise.reject(new Error('Pending action'));
        dispatch(pending(FETCH_BY_ID));
        return movieService.read(id)
            .then((res) => {
                dispatch(fetchMovieSuccess(res.data));
            })
            .catch((e) => {
                const error = e?.response?.data.messages
                    ? new Error(e?.response?.data.messages) : e;
                dispatch(handleError(error.message));
                throw error;
            });
    };
}

export function fetchMoviesPagination(sortBy, filter, search, offset) {
    return (dispatch, getState) => {
        if (getState().movies.pending) return Promise.reject(new Error('Pending action'));
        dispatch(pending(FETCH_BY_SEARCH));
        return movieService.searchMovies(sortBy, filter, search, offset)
            .then((res) => {
                dispatch(fetchMoviesPaginationSuccess(res.data));
            })
            .catch((e) => {
                const error = e?.response?.data.messages
                    ? new Error(e?.response?.data.messages) : e;
                dispatch(handleError(error.message));
                throw error;
            });
    };
}

export function fetchMoviesIfNeed() {
    return (dispatch, getState) => {
        const { movies } = getState();
        if (movies || movies === []) {
            return dispatch(fetchMovies());
        } else {
            return Promise.resolve();
        }
    };
}

function addMovieSuccess(data) {
    return dispatch => dispatch({
        type: ADD_MOVIE_SUCCESS,
        payload: data,
    });
}

export function addMovie(data) {
    return (dispatch, getState) => {
        if (getState().movies.pending) return Promise.reject(new Error('Pending action'));
        dispatch(pending());
        return movieService.create(data)
            .then((res) => {
                dispatch(addMovieSuccess(res.data));
            })
            .catch((e) => {
                const error = e?.response?.data.messages
                    ? new Error(e?.response?.data.messages) : e;
                dispatch(handleError(error.message));
                throw error;
            });
    };
}

function editMovieSuccess(data) {
    return dispatch => dispatch({
        type: EDIT_MOVIE_SUCCESS,
        payload: data,
    });
}

export function editMovie(data) {
    return (dispatch, getState) => {
        if (getState().movies.pending) return Promise.reject(new Error('Pending action'));
        dispatch(pending());
        return movieService.update(data)
            .then((res) => {
                dispatch(editMovieSuccess(res.data));
            })
            .catch((e) => {
                const error = e?.response?.data.messages
                    ? new Error(e?.response?.data.messages) : e;
                dispatch(handleError(error.message));
                throw error;
            });
    };
}

export function deleteMovieSuccess(movieId) {
    return dispatch => dispatch({
        type: DELETE_MOVIE_SUCCESS,
        payload: movieId,
    });
}

export function deleteMovie(id) {
    return (dispatch, getState) => {
        if (getState().movies.pending) return Promise.reject(new Error('Pending action'));
        dispatch(pending());
        return movieService.remove(id)
            .then(() => {
                dispatch(deleteMovieSuccess(id));
            })
            .catch((e) => {
                const error = e?.response?.data.messages
                    ? new Error(e?.response?.data.messages) : e;
                dispatch(handleError(error.message));
                throw error;
            });
    };
}
