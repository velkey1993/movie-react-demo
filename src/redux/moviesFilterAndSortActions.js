export const FILTER_MOVIES_BY_GENRE = 'FILTER_MOVIES_BY_GENRE';
export const SORT_MOVIES_BY_TYPE = 'SORT_MOVIES_BY_TYPE';

export const filterMoviesByGenre = genre => dispatch => dispatch({
    type: 'FILTER_MOVIES_BY_GENRE',
    payload: genre,
});

export const sortMoviesByType = type => dispatch => dispatch({
    type: 'SORT_MOVIES_BY_TYPE',
    payload: type,
});
