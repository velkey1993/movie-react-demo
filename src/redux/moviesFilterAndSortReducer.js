const initialState = {
    genreFilter: null,
    sortType: 'genre',
};

const reducer = (state = initialState, action) => {
    switch (action.type) {
        case 'FILTER_MOVIES_BY_GENRE':
            return {
                ...state,
                genreFilter: action.payload,
            };
        case 'SORT_MOVIES_BY_TYPE':
            return {
                ...state,
                sortType: action.payload,
            };
        default:
            return state;
    }
};

export default reducer;
