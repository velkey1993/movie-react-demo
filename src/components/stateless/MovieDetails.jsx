import React from 'react';
import './MovieDetails.css';
import { Button, Col, Row } from 'react-bootstrap';

function MovieDetails({ movie, closeDetails }) {
    return (
        <div id='movie-details-container' className='jumbotron'>
            <div id='movie-details-container-background' />
            <div id='movie-details-container-content'>
                <div id='movie-details-container-data'>
                    <Row>
                        <Col xs={11} sm={11} md={11} lg={11} xl={11}>
                            <p id='movie-details-container-page-name'>
                                <b>epam</b>
                                roulette
                            </p>
                        </Col>
                        <Col xs={1} sm={1} md={1} lg={1} xl={1}>
                            <div>
                                <Button
                                    id='movie-details-container-search-button'
                                    variant='primary'
                                    onClick={closeDetails}
                                >
                                    <span className='glyphicon glyphicon-search' />
                                </Button>
                            </div>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={6} sm={5} md={4} lg={3} xl={1}>
                            <img src={movie.poster_path} alt={movie.title} />
                        </Col>
                        <Col xs={6} sm={5} md={6} lg={7} xl={9}>
                            <Row>
                                <Col xs={12} sm={12} md={12} lg={12} xl={12}>
                                    <p id='movie-details-container-data-title'>
                                        {movie.title}
                                        &emsp;
                                        <span id='movie-details-container-data-vote-average'>
                                            {movie.vote_average?.toFixed(1) || Number.NaN}
                                        </span>
                                    </p>
                                </Col>
                            </Row>
                            <Row>
                                <Col xs={12} sm={12} md={12} lg={12} xl={12}>
                                    <p>{movie.tagline}</p>
                                </Col>
                            </Row>
                            <Row>
                                <Col xs={12} sm={12} md={12} lg={12} xl={12}>
                                    <p style={{ color: '#f65261' }}>
                                        {new Date(movie.release_date).toISOString().slice(0, 4)}
                                        &emsp;
                                        {movie.runtime}
                                        {' '}
                                        min
                                    </p>
                                </Col>
                            </Row>
                            <Row>
                                <Col xs={12} sm={12} md={12} lg={12} xl={12}>
                                    <p style={{ fontSize: '120%' }}>{movie.overview}</p>
                                </Col>
                            </Row>
                        </Col>
                    </Row>
                </div>
            </div>

        </div>
    );
}

export default MovieDetails;
