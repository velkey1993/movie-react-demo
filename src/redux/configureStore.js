import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import moviesReducer from './moviesReducer';
import moviesFilterAndSortReducer from './moviesFilterAndSortReducer';

const rootReducer = combineReducers({
    movies: moviesReducer,
    filterAndSort: moviesFilterAndSortReducer,
});

export default function configureStore(preloadedState) {
    return createStore(
        rootReducer,
        preloadedState,
        applyMiddleware(thunkMiddleware),
    );
}
