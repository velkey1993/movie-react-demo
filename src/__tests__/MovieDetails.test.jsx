import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import configureStore from 'redux-mock-store';
import * as ReactRedux from 'react-redux';
import MovieDetails from '../components/stateless/MovieDetails';

const PLACEHOLDER_TEXT = 'Test Placeholder';

test('should contains placeholder text if movie is not present', () => {
    const { getByText } = render(
        <Provider store={createStore(() => { })}>
            <MovieDetails placeholder={PLACEHOLDER_TEXT} />
        </Provider>,
    );
    const element = getByText(new RegExp(PLACEHOLDER_TEXT, 'i'));
    expect(element).toBeInTheDocument();
});

test('should not contains placeholder text if movie present', () => {
    const movieMock = {
        title: 'Test title',
        release_date: '2017-12-01',
        poster_path: 'https://image.tmdb.org/t/p/w500/k4FwHlMhuRR5BISY2Gm2QZHlH5Q.jpg',
        genres: ['Drama', 'Fantasy', 'Romance'],
        overview: 'An other-worldly story ...',
        runtime: 123,
    };
    const { queryByText } = render(
        <Provider store={createStore(() => { })}>
            <MovieDetails movie={movieMock} placeholder={PLACEHOLDER_TEXT} />
        </Provider>,
    );
    const element = queryByText(new RegExp(PLACEHOLDER_TEXT, 'i'));
    expect(element).not.toBeInTheDocument();
});

test('should show movie title', () => {
    const movieMock = {
        title: 'Test title',
        release_date: '2017-12-01',
        poster_path: 'https://image.tmdb.org/t/p/w500/k4FwHlMhuRR5BISY2Gm2QZHlH5Q.jpg',
        genres: ['Drama', 'Fantasy', 'Romance'],
        overview: 'An other-worldly story ...',
        runtime: 123,
    };
    const { getByText } = render(
        <Provider store={createStore(() => { })}>
            <MovieDetails movie={movieMock} placeholder={PLACEHOLDER_TEXT} />
        </Provider>,
    );
    const element = getByText(new RegExp(movieMock.title, 'i'));
    expect(element).toBeInTheDocument();
});

test('renders correctly', () => {
    const store = configureStore()({});
    jest
        .spyOn(ReactRedux, 'useDispatch')
        .mockImplementation(() => store.dispatch);

    const movieMock = {
        title: 'Test title',
        release_date: '2017-12-01',
        poster_path: 'https://image.tmdb.org/t/p/w500/k4FwHlMhuRR5BISY2Gm2QZHlH5Q.jpg',
        genres: ['Drama', 'Fantasy', 'Romance'],
        overview: 'An other-worldly story ...',
        runtime: 123,
    };
    const testView = shallow(
        <MovieDetails movie={movieMock} placeholder={PLACEHOLDER_TEXT} />,
    );
    expect(testView).toMatchSnapshot();
});
