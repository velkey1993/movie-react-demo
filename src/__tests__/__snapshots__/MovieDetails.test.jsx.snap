// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="jumbotron"
  id="movie-details-container"
>
  <div
    id="movie-details-container-background"
  />
  <div
    id="movie-details-container-content"
  >
    <div
      id="movie-details-container-data"
    >
      <Row
        noGutters={false}
      >
        <Col
          lg={11}
          md={11}
          sm={11}
          xl={11}
          xs={11}
        >
          <p
            id="movie-details-container-page-name"
          >
            <b>
              epam
            </b>
            roulette
          </p>
        </Col>
        <Col
          lg={1}
          md={1}
          sm={1}
          xl={1}
          xs={1}
        >
          <div>
            <button
              className="btn btn-primary"
              id="movie-details-container-search-button"
              onClick={[Function]}
              type="button"
            >
              <span
                className="glyphicon glyphicon-search"
              />
            </button>
          </div>
        </Col>
      </Row>
      <Row
        noGutters={false}
      >
        <Col
          lg={3}
          md={4}
          sm={5}
          xl={1}
          xs={6}
        >
          <img
            alt="Test title"
            src="https://image.tmdb.org/t/p/w500/k4FwHlMhuRR5BISY2Gm2QZHlH5Q.jpg"
          />
        </Col>
        <Col
          lg={7}
          md={6}
          sm={5}
          xl={9}
          xs={6}
        >
          <Row
            noGutters={false}
          >
            <Col
              lg={12}
              md={12}
              sm={12}
              xl={12}
              xs={12}
            >
              <p
                id="movie-details-container-data-title"
              >
                Test title
                 
                <span
                  id="movie-details-container-data-vote-average"
                >
                  NaN
                </span>
              </p>
            </Col>
          </Row>
          <Row
            noGutters={false}
          >
            <Col
              lg={12}
              md={12}
              sm={12}
              xl={12}
              xs={12}
            >
              <p />
            </Col>
          </Row>
          <Row
            noGutters={false}
          >
            <Col
              lg={12}
              md={12}
              sm={12}
              xl={12}
              xs={12}
            >
              <p
                style={
                  Object {
                    "color": "#f65261",
                  }
                }
              >
                2017
                 
                123
                 
                min
              </p>
            </Col>
          </Row>
          <Row
            noGutters={false}
          >
            <Col
              lg={12}
              md={12}
              sm={12}
              xl={12}
              xs={12}
            >
              <p
                style={
                  Object {
                    "fontSize": "120%",
                  }
                }
              >
                An other-worldly story ...
              </p>
            </Col>
          </Row>
        </Col>
      </Row>
    </div>
  </div>
</div>
`;
